Build zlib from source for Windows wheels

NeoKish opened the pull request:
<!--
Thanks for contributing to h5py!

Before opening a pull request, please:

- Run simple static checks with `tox -e pre-commit`
- Run the tests with e.g. `tox -e py312-test-deps`
- If your change is visible to someone using or building h5py, add a release
  note in the news/ folder.

For more information, see the contribution guide:
http://docs.h5py.org/en/stable/contributing.html#how-to-get-your-code-into-h5py

-->

- Fixes #2261 
- Script for building zlib from source is added instead of nuget installation(which is at lower version)

takluyver commented on ci/cibw_before_all_windows.sh:
```diff
@@ -23,9 +24,25 @@ if [[ "$ARCH" == "ARM64" ]]; then
-    ZLIB_ROOT="$PROJECT_PATH/zlib-msvc-x64/build/native"
+    # Build zlib from source for Windows AMD64
+    mkdir zlib_win_x64
+    curl -sLO https://github.com/madler/zlib/archive/refs/tags/v$ZLIB_VERSION.tar.gz
```
For Mac, we download a tarball from `https://zlib.net/fossils/zlib-$ZLIB_VERSION.tar.gz`. Does the archive from Github give us anything extra? If not, could we use the same URL on Windows.

takluyver commented on ci/cibw_before_all_windows.sh:
```diff
@@ -23,9 +24,25 @@ if [[ "$ARCH" == "ARM64" ]]; then
+    cmake --build build --config Debug 
+    cmake --install build --config Debug
+    mv zlib_win_x64/bin zlib-win-x64/bin_debug
+    mv zlib_win_x64/lib zlib-win-x64/lib_debug
```
Is there any reason to produce a debug build, if we're only going to use the release build?

takluyver commented on ci/cibw_before_all_windows.sh:
```diff
@@ -23,9 +24,25 @@ if [[ "$ARCH" == "ARM64" ]]; then
+    # Release build
+    cmake --build build --config Release
+    cmake --install build --config Release
+    mv zlib_win_x64/bin zlib-win-x64/bin_release
```
```
mv: cannot move 'zlib_win_x64/bin' to 'zlib-win-x64/bin_release': No such file or directory
```

On the Windows AMD64 job.

takluyver reviewed (commented):
Thanks, a few things to figure out.

NeoKish commented:
@takluyver , thanks for the suggestions. 

- I have added the separated zlib script for use in both cibw_before_all_windows.sh and azure-pipelines-steps.yml and updated the paths accordingly

NeoKish commented on ci/fix_paths.py:
```diff
@@ -29,7 +29,7 @@ def main():
                 f = pjoin(zlib_root, 'bin', 'zlib1.dll')
             elif arch in ("amd64", "x86_64"):
-                f = pjoin(zlib_root, 'bin_release', 'zlib.dll')
+                f = pjoin(zlib_root, 'bin_release', 'zlib1.dll')
```
Have modified this

takluyver commented:
Thanks! One more thing: the Github workflow (`.github/workflows/build_wheels.yml`) still calls an action to setup nuget, although we're no longer using it anywhere, as far as I can see. Could you get rid of that?
